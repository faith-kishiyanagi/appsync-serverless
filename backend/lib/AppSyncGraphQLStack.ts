import * as cdk from '@aws-cdk/core';
import * as appsync from '@aws-cdk/aws-appsync';
import * as lambda from '@aws-cdk/aws-lambda';

export class AppSyncGraphQLStack extends cdk.Stack {

  readonly api: appsync.GraphqlApi;

  constructor(scope: cdk.Construct, id: string, resolver: lambda.Function, props?: cdk.StackProps) {
    super(scope, id, props);

    this.api = new appsync.GraphqlApi(this, 'GqlApi', {
      name: 'blog-graphql-api',
      schema: appsync.Schema.fromAsset('graphql/schema.graphql'),
      authorizationConfig: {
        defaultAuthorization: {
          authorizationType: appsync.AuthorizationType.API_KEY,
          apiKeyConfig: {
            expires: cdk.Expiration.after(cdk.Duration.days(2)),
          }
        }
      }
    })

    const postResolver = this.api.addLambdaDataSource('PostResolver', resolver)

    postResolver.createResolver({
      typeName: 'Query',
      fieldName: 'allPosts'
    });
    postResolver.createResolver({
      typeName: 'Query',
      fieldName: 'findOnePostBy'
    });
    postResolver.createResolver({
      typeName: 'Mutation',
      fieldName: 'createPost'
    });
    postResolver.createResolver({
      typeName: 'Mutation',
      fieldName: 'updatePost'
    });
    postResolver.createResolver({
      typeName: 'Mutation',
      fieldName: 'deletePost'
    });

    new cdk.CfnOutput(this, 'AppSyncUrl', {
      value: this.api.graphqlUrl
    });
    new cdk.CfnOutput(this, 'AppSyncApiKey', {
      value: this.api.apiKey || ''
    });
    new cdk.CfnOutput(this, 'ProjectRegion', {
      value: this.region
    });
  }
}
