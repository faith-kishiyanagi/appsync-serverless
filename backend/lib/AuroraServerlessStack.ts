import * as cdk from '@aws-cdk/core';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as rds from '@aws-cdk/aws-rds';
import { AuroraCapacityUnit } from '@aws-cdk/aws-rds';

export class AuroraServerlessStack extends cdk.Stack {

  readonly vpc: ec2.Vpc;
  readonly blogDB: rds.ServerlessCluster;

  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    this.vpc = new ec2.Vpc(this, 'BlogAppVpc', {
      cidr: '10.0.1.0/24',
      natGateways: 0,
      subnetConfiguration: [
        {cidrMask: 26, name: 'Isolated', subnetType: ec2.SubnetType.ISOLATED}
      ],
    })

    this.blogDB = new rds.ServerlessCluster(this, 'BlogDbAuroraServerlessCluster', {
      engine: rds.DatabaseClusterEngine.AURORA_MYSQL,
      parameterGroup: rds.ParameterGroup.fromParameterGroupName(this, 'BlogDbParameterGroup', 'default.aurora-mysql5.7'),
      defaultDatabaseName: 'BlogDB',
      vpc: this.vpc,
      vpcSubnets: {subnetType: ec2.SubnetType.ISOLATED},
      scaling: {
        autoPause: cdk.Duration.minutes(5),
        minCapacity: AuroraCapacityUnit.ACU_1,
        maxCapacity: AuroraCapacityUnit.ACU_2,
      }
    })
  }
}
