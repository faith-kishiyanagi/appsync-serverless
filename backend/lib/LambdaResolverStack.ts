import * as cdk from '@aws-cdk/core';
import * as lambda from '@aws-cdk/aws-lambda';
import * as rds from '@aws-cdk/aws-rds';

export class LambdaResolverStack extends cdk.Stack {

  readonly postFunction: lambda.Function;

  constructor(scope: cdk.Construct, id: string, db: rds.ServerlessCluster, props?: cdk.StackProps) {
    super(scope, id, props);

    this.postFunction = new lambda.Function(this, 'PostFunction', {
      runtime: lambda.Runtime.NODEJS_14_X,
      code: new lambda.AssetCode('functions/post'),
      handler: 'src/index.handler',
      memorySize: 256,
      timeout: cdk.Duration.seconds(30),
      environment: {
        CLUSTER_ARN: db.clusterArn,
        SECRET_ARN: db.secret?.secretArn || '',
        DB_NAME: 'BlogDB',
        AWS_NODEJS_CONNECTION_REUSE_ENABLED: '1'
      }
    })

    db.grantDataApiAccess(this.postFunction);
  }
}
