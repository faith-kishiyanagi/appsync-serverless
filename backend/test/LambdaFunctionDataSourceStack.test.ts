import '@aws-cdk/assert/jest'
import { SynthUtils } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import { LambdaResolverStack } from '../lib/LambdaResolverStack';
import { AuroraServerlessStack } from "../lib/AuroraServerlessStack";

const app = new cdk.App();
const aurora = new AuroraServerlessStack(app, 'AuroraServerless');
const stack = new LambdaResolverStack(app, 'LambdaFunctionDataSource', aurora.blogDB);

test('Stack snapshot', () => {
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

test('Lambda Function', () => {
  expect(stack).toHaveResource('AWS::Lambda::Function');
});
