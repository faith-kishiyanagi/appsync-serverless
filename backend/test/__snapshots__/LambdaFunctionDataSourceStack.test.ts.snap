// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack snapshot 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters91625822e8bbab3db208ace6122c9a01d6f5e24184becf2bec9891d8ac9234eeArtifactHash5F594A89": Object {
      "Description": "Artifact hash for asset \\"91625822e8bbab3db208ace6122c9a01d6f5e24184becf2bec9891d8ac9234ee\\"",
      "Type": "String",
    },
    "AssetParameters91625822e8bbab3db208ace6122c9a01d6f5e24184becf2bec9891d8ac9234eeS3Bucket4994A5FA": Object {
      "Description": "S3 bucket for asset \\"91625822e8bbab3db208ace6122c9a01d6f5e24184becf2bec9891d8ac9234ee\\"",
      "Type": "String",
    },
    "AssetParameters91625822e8bbab3db208ace6122c9a01d6f5e24184becf2bec9891d8ac9234eeS3VersionKey556059D3": Object {
      "Description": "S3 key for asset version \\"91625822e8bbab3db208ace6122c9a01d6f5e24184becf2bec9891d8ac9234ee\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "PostFunction61046647": Object {
      "DependsOn": Array [
        "PostFunctionServiceRoleDefaultPolicy2A61B0F2",
        "PostFunctionServiceRole1BB2A70F",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters91625822e8bbab3db208ace6122c9a01d6f5e24184becf2bec9891d8ac9234eeS3Bucket4994A5FA",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters91625822e8bbab3db208ace6122c9a01d6f5e24184becf2bec9891d8ac9234eeS3VersionKey556059D3",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters91625822e8bbab3db208ace6122c9a01d6f5e24184becf2bec9891d8ac9234eeS3VersionKey556059D3",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "CLUSTER_ARN": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "arn:",
                  Object {
                    "Ref": "AWS::Partition",
                  },
                  ":rds:",
                  Object {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  Object {
                    "Ref": "AWS::AccountId",
                  },
                  ":cluster:",
                  Object {
                    "Fn::ImportValue": "AuroraServerless:ExportsOutputRefBlogDbAuroraServerlessClusterB47832F535767072",
                  },
                ],
              ],
            },
            "DB_NAME": "BlogDB",
            "SECRET_ARN": Object {
              "Fn::ImportValue": "AuroraServerless:ExportsOutputRefBlogDbAuroraServerlessClusterSecretAttachment26906F9D1FB8943F",
            },
          },
        },
        "Handler": "src/index.handler",
        "MemorySize": 256,
        "Role": Object {
          "Fn::GetAtt": Array [
            "PostFunctionServiceRole1BB2A70F",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "PostFunctionServiceRole1BB2A70F": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "PostFunctionServiceRoleDefaultPolicy2A61B0F2": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "rds-data:BatchExecuteStatement",
                "rds-data:BeginTransaction",
                "rds-data:CommitTransaction",
                "rds-data:ExecuteStatement",
                "rds-data:RollbackTransaction",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "secretsmanager:GetSecretValue",
                "secretsmanager:DescribeSecret",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::ImportValue": "AuroraServerless:ExportsOutputRefBlogDbAuroraServerlessClusterSecretAttachment26906F9D1FB8943F",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PostFunctionServiceRoleDefaultPolicy2A61B0F2",
        "Roles": Array [
          Object {
            "Ref": "PostFunctionServiceRole1BB2A70F",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
